File: src/lib/dbscan.ts
export interface DBSCANInputPoint {
  id: string;
  features: number[]; // [hue, saturation, value, circularity, aspectRatio]
  [key: string]: any;
}

export interface DBSCANResult {
  clusterId: number; // -1 for noise, 0, 1, 2... for clusters
  isCore: boolean;
  isBorder: boolean;
  isNoise: boolean;
}

// Normalize all features of a list of points to a [0, 1] range.
// Hue (index 0) is circular [0, 360] — we encode it as sin/cos pair so
// that hue 355° and 5° are correctly treated as close.
export function normalizeFeatures(points: DBSCANInputPoint[]): DBSCANInputPoint[] {
  if (points.length === 0) return [];

  const numFeatures = points[0].features.length; // [hue, sat, val, circ, ar]

  // For non-hue features find min/max for min-max scaling
  const mins = new Array(numFeatures).fill(Infinity);
  const maxs = new Array(numFeatures).fill(-Infinity);

  for (const p of points) {
    for (let i = 1; i < numFeatures; i++) { // skip index 0 (hue)
      const v = p.features[i];
      if (v < mins[i]) mins[i] = v;
      if (v > maxs[i]) maxs[i] = v;
    }
  }

  // Expand hue into sin/cos components, then min-max normalize the rest.
  // Final feature vector: [sinH_norm, cosH_norm, sat_norm, val_norm, circ_norm, ar_norm]
  return points.map(p => {
    const hueRad = (p.features[0] / 360) * 2 * Math.PI;
    // sin/cos in [-1,1] → shift to [0,1]
    const sinH = (Math.sin(hueRad) + 1) / 2;
    const cosH = (Math.cos(hueRad) + 1) / 2;

    const rest = p.features.slice(1).map((v, i) => {
      const fi = i + 1; // original feature index
      const range = maxs[fi] - mins[fi];
      if (range === 0) return 0.5;
      return (v - mins[fi]) / range;
    });

    return {
      ...p,
      features: [sinH, cosH, ...rest],
      rawFeatures: [...p.features],
    };
  });
}

// Compute weighted Euclidean distance between two feature vectors
export function euclideanDistance(a: number[], b: number[], weights: number[]): number {
  let sum = 0;
  let weightSum = 0;
  for (let i = 0; i < a.length; i++) {
    const w = weights[i] !== undefined ? weights[i] : 1;
    sum += w * Math.pow(a[i] - b[i], 2);
    weightSum += w;
  }
  return Math.sqrt(sum / (weightSum || 1));
}

// Build weight vector for the expanded feature space [sinH, cosH, sat, val, circ, ar]
export function buildWeights(featureMode: string): number[] {
  if (featureMode === "color") {
    // Emphasise hue (both sin+cos) and saturation, ignore shape
    return [1.5, 1.5, 1.0, 0.5, 0, 0];
  } else if (featureMode === "shape") {
    // Ignore color, emphasise circularity and aspect ratio
    return [0, 0, 0, 0, 1.5, 1.0];
  }
  // "both" — balanced
  return [1.0, 1.0, 0.8, 0.8, 1.2, 1.0];
}

// Run DBSCAN
export function runDBSCAN(
  points: DBSCANInputPoint[],
  eps: number,
  minSamples: number,
  weights: number[]
): Map<string, DBSCANResult> {
  const n = points.length;
  const results = new Map<string, DBSCANResult>();

  // Initialize all as noise
  for (const p of points) {
    results.set(p.id, { clusterId: -1, isCore: false, isBorder: false, isNoise: true });
  }

  if (n === 0) return results;

  const getNeighbors = (idx: number): number[] => {
    const neighbors: number[] = [];
    const p1 = points[idx];
    for (let i = 0; i < n; i++) {
      if (euclideanDistance(p1.features, points[i].features, weights) <= eps) {
        neighbors.push(i);
      }
    }
    return neighbors;
  };

  const visited = new Set<number>();
  let currentClusterId = 0;

  for (let i = 0; i < n; i++) {
    if (visited.has(i)) continue;
    visited.add(i);

    const neighbors = getNeighbors(i);
    if (neighbors.length < minSamples) continue; // stays noise

    // Core point
    results.set(points[i].id, { clusterId: currentClusterId, isCore: true, isBorder: false, isNoise: false });

    const queue: number[] = neighbors.filter(idx => idx !== i);

    for (let qIdx = 0; qIdx < queue.length; qIdx++) {
      const neighborIdx = queue[qIdx];

      if (!visited.has(neighborIdx)) {
        visited.add(neighborIdx);
        const nn = getNeighbors(neighborIdx);

        if (nn.length >= minSamples) {
          for (const nnIdx of nn) {
            if (!queue.includes(nnIdx) && nnIdx !== i) queue.push(nnIdx);
          }
          results.set(points[neighborIdx].id, { clusterId: currentClusterId, isCore: true, isBorder: false, isNoise: false });
        } else {
          results.set(points[neighborIdx].id, { clusterId: currentClusterId, isCore: false, isBorder: true, isNoise: false });
        }
      } else {
        const curr = results.get(points[neighborIdx].id)!;
        if (curr.clusterId === -1) {
          results.set(points[neighborIdx].id, { ...curr, clusterId: currentClusterId, isNoise: false, isBorder: !curr.isCore });
        }
      }
    }

    currentClusterId++;
  }

  return results;
}

// --- Nearest-neighbour fallback classifier ---
// When a point is classified as noise by DBSCAN, find the closest
// non-noise point and adopt its cluster so the user always gets a
// meaningful fruit label (not "Derau").
export function assignNearestCluster(
  noisePointId: string,
  allPoints: DBSCANInputPoint[],
  results: Map<string, DBSCANResult>,
  weights: number[]
): DBSCANResult {
  const noisePoint = allPoints.find(p => p.id === noisePointId);
  if (!noisePoint) return { clusterId: -1, isCore: false, isBorder: false, isNoise: true };

  let bestDist = Infinity;
  let bestResult: DBSCANResult = { clusterId: -1, isCore: false, isBorder: false, isNoise: true };

  for (const p of allPoints) {
    if (p.id === noisePointId) continue;
    const r = results.get(p.id);
    if (!r || r.clusterId === -1) continue; // skip other noise points

    const dist = euclideanDistance(noisePoint.features, p.features, weights);
    if (dist < bestDist) {
      bestDist = dist;
      bestResult = { clusterId: r.clusterId, isCore: false, isBorder: true, isNoise: false };
    }
  }

  return bestResult;
}

File: src/lib/dbscan.py
import sys
import json
import math

def circular_hue_distance(h1, h2):
    d = abs(h1 - h2) % 360
    return 360 - d if d > 180 else d

def normalize_features(points):
    if not points:
        return []
    
    n_points = len(points)
    # Features: [hue, sat, val, circ, ar]
    num_features = len(points[0]['features'])
    
    # Calculate min-max for non-hue features (index 1 to end)
    mins = [float('inf')] * num_features
    maxs = [float('-inf')] * num_features
    
    for p in points:
        for i in range(1, num_features):
            v = p['features'][i]
            if v < mins[i]:
                mins[i] = v
            if v > maxs[i]:
                maxs[i] = v
                
    normalized = []
    for p in points:
        hue = p['features'][0]
        # Encode hue circularity using sin and cos shifted to [0, 1]
        hue_rad = (hue / 360.0) * 2.0 * math.pi
        sin_h = (math.sin(hue_rad) + 1.0) / 2.0
        cos_h = (math.cos(hue_rad) + 1.0) / 2.0
        
        rest = []
        for i in range(1, num_features):
            range_val = maxs[i] - mins[i]
            if range_val == 0:
                rest.append(0.5)
            else:
                rest.append((p['features'][i] - mins[i]) / range_val)
                
        normalized.append({
            'id': p['id'],
            'features': [sin_h, cos_h] + rest,
            'raw_features': p['features']
        })
        
    return normalized

def euclidean_distance(a, b, weights):
    sum_sq = 0.0
    weight_sum = 0.0
    for i in range(len(a)):
        w = weights[i] if i < len(weights) else 1.0
        sum_sq += w * ((a[i] - b[i]) ** 2)
        weight_sum += w
    return math.sqrt(sum_sq / (weight_sum or 1.0))

def run_dbscan(points, eps, min_samples, weights):
    n = len(points)
    results = {
        p['id']: {
            'clusterId': -1,
            'isCore': False,
            'isBorder': False,
            'isNoise': True
        } for p in points
    }
    
    if n == 0:
        return results
        
    def get_neighbors(idx):
        neighbors = []
        p1 = points[idx]
        for i in range(n):
            if euclidean_distance(p1['features'], points[i]['features'], weights) <= eps:
                neighbors.append(i)
        return neighbors

    visited = set()
    current_cluster_id = 0
    
    for i in range(n):
        if i in visited:
            continue
        visited.add(i)
        
        neighbors = get_neighbors(i)
        if len(neighbors) < min_samples:
            continue
            
        # Core point found
        results[points[i]['id']] = {
            'clusterId': current_cluster_id,
            'isCore': True,
            'isBorder': False,
            'isNoise': False
        }
        
        queue = [idx for idx in neighbors if idx != i]
        q_idx = 0
        while q_idx < len(queue):
            neighbor_idx = queue[q_idx]
            
            if neighbor_idx not in visited:
                visited.add(neighbor_idx)
                nn = get_neighbors(neighbor_idx)
                
                if len(nn) >= min_samples:
                    for nn_idx in nn:
                        if nn_idx not in queue and nn_idx != i:
                            queue.append(nn_idx)
                    results[points[neighbor_idx]['id']] = {
                        'clusterId': current_cluster_id,
                        'isCore': True,
                        'isBorder': False,
                        'isNoise': False
                    }
                else:
                    results[points[neighbor_idx]['id']] = {
                        'clusterId': current_cluster_id,
                        'isCore': False,
                        'isBorder': True,
                        'isNoise': False
                    }
            else:
                curr = results[points[neighbor_idx]['id']]
                if curr['clusterId'] == -1:
                    results[points[neighbor_idx]['id']] = {
                        'clusterId': current_cluster_id,
                        'isCore': curr['isCore'],
                        'isBorder': not curr['isCore'],
                        'isNoise': False
                    }
            q_idx += 1
            
        current_cluster_id += 1
        
    return results

def main():
    try:
        input_data = json.load(sys.stdin)
        points = input_data.get('points', [])
        eps = input_data.get('eps', 0.35)
        min_samples = input_data.get('minSamples', 3)
        weights = input_data.get('weights', [1.0, 1.0, 1.0, 1.0, 1.0, 1.0])
        
        normalized = normalize_features(points)
        results = run_dbscan(normalized, eps, min_samples, weights)
        
        print(json.dumps(results))
    except Exception as e:
        sys.stderr.write(str(e) + "\n")
        sys.exit(1)

if __name__ == '__main__':
    main()
